### ISSUE TITLE
'Namespace' object has no attribute 'certbot_dns_cpanel:cpanel_credentials'.

### ISSUE THREAD

**Issue by Alir3z4** [badjware/certbot-dns-cpanel#9]
Thanks for the great work on this.

I've made the cert by running:
```
certbot --authenticator certbot-dns-cpanel:cpanel --installer nginx --certbot-dns-cpanel:cpanel-credentials /etc/letsencrypt/cpanel_dns_credentials.ini --agree-tos -w /var/www/certbot --email ali@domain.com -d domain.com -d *.domain.com --force-renewal
```

I've experienced an error while trying to run the `certbot renew`

```
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /etc/letsencrypt/renewal/domain.com.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Renewal configuration file /etc/letsencrypt/renewal/domain.com.conf (cert: domain.com) produced an unexpected error: 'Namespace' object has no attribute 'certbot_dns_cpanel:cpanel_credentials'. Skipping.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

No renewals were attempted.

Additionally, the following renewal configurations were invalid:
  /etc/letsencrypt/renewal/domain.com.conf (parsefail)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
0 renew failure(s), 1 parse failure(s)
````
Content of `/etc/letsencrypt/renewal/domain.com.conf`

```
root@7714b64abcc2:/# cat /etc/letsencrypt/renewal/domain.com.conf
# renew_before_expiry = 30 days
version = 0.31.0
archive_dir = /etc/letsencrypt/archive/domain.com
cert = /etc/letsencrypt/live/domain.com/cert.pem
privkey = /etc/letsencrypt/live/domain.com/privkey.pem
chain = /etc/letsencrypt/live/domain.com/chain.pem
fullchain = /etc/letsencrypt/live/domain.com/fullchain.pem

# Options used in the renewal process
[renewalparams]
account = xxxxxxxxxxxxxxx
authenticator = certbot-dns-cpanel:cpanel
webroot_path = /var/www/certbot,
server = https://acme-v02.api.letsencrypt.org/directory
installer = nginx
certbot_dns_cpanel:cpanel_credentials = /etc/letsencrypt/cpanel_dns_credentials.ini
[[webroot_map]]
```

Content of log `/var/log/letsencrypt/letsencrypt.log`
```
root@7714b64abcc2:/# cat /var/log/letsencrypt/letsencrypt.log
2020-06-21 13:49:17,427:DEBUG:certbot.main:certbot version: 0.31.0
2020-06-21 13:49:17,427:DEBUG:certbot.main:Arguments: []
2020-06-21 13:49:17,428:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#manual,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2020-06-21 13:49:17,444:DEBUG:certbot.log:Root logging level set at 20
2020-06-21 13:49:17,444:INFO:certbot.log:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2020-06-21 13:49:17,458:DEBUG:certbot.plugins.selection:Requested authenticator <certbot.cli._Default object at 0x7f0ed022bd68> and installer <certbot.cli._Default object at 0x7f0ed022bd68>
2020-06-21 13:49:17,458:WARNING:certbot.renewal:Renewal configuration file /etc/letsencrypt/renewal/domain.com.conf (cert: domain.com) produced an unexpected error: 'Namespace' object has no attribute 'certbot_dns_cpanel:cpanel_credentials'. Skipping.
2020-06-21 13:49:17,463:DEBUG:certbot.renewal:Traceback was:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 419, in handle_renewal_request
    renewal_candidate = _reconstitute(lineage_config, renewal_file)
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 86, in _reconstitute
    _restore_plugin_configs(config, renewalparams)
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 154, in _restore_plugin_configs
    if config_item.startswith(plugin_prefix + "_") and not cli.set_by_cli(config_item):
  File "/usr/lib/python3/dist-packages/certbot/cli.py", line 219, in set_by_cli
    if not isinstance(getattr(detector, var), _Default):
AttributeError: 'Namespace' object has no attribute 'certbot_dns_cpanel:cpanel_credentials'

2020-06-21 13:49:17,464:DEBUG:certbot.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/certbot", line 11, in <module>
    load_entry_point('certbot==0.31.0', 'console_scripts', 'certbot')()
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1365, in main
    return config.func(config, plugins)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1272, in renew
    renewal.handle_renewal_request(config)
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 477, in handle_renewal_request
    len(renew_failures), len(parse_failures)))
certbot.errors.Error: 0 renew failure(s), 1 parse failure(s)
```

The machine is a Debian 10 (_in docker_)

Versions:
```
root@7714b64abcc2:/# python3 --version
Python 3.7.3
root@7714b64abcc2:/# certbot --version
certbot 0.31.0
```

**Comment by badjware** [comment]
This is very strange, I did not manage to reproduce your issue on my side yet, but looking through your post I see the following output:
```
Additionally, the following renewal configurations were invalid:
  /etc/letsencrypt/renewal/domain.com.conf (parsefail)
```

Could you please validate the content of this file? The content that you pasted looks okay, but just double-check to be sure.

**Comment by Alir3z4** [comment]
Checked through the source code of bunch of other plugins and also development materials for implementing certbot plugins but couldn't find anything much supporting 

```
'Namespace' object has no attribute ...
```

This is mostly a generic error on Python AttributeError.
Anyway, I've found out the issue is file ownership of  `/etc/letsencrypt/cpanel_dns_credentials.ini`.

The directory of Nginx config and Let's Encrypt are created on host machine and mounted to the container running. I've created the file `/etc/letsencrypt/cpanel_dns_credentials.ini` via the host machine and then certbot was accessing via the container itself and due to differences in user/uid of file ownership, the certbot was raising that error.

Running `chown root:root /etc/letsencrypt/cpanel_dns_credentials.ini` within the container fixed the problem.

_I guess this issue could be a good piece of information to be mentioned on the FAQ or README file if people are running things via container, or maybe I'm simply a noob in docker file permission problems 🤷_
